Implement traffic accounting and limiting via an eBPF module

f00b4r0 opened the pull request:
Opening this PR for easy reference until the relevant changes are fully tested and merged

This PR addresses #9 and #25 implementing support for `Acct-{Input,Output}-{Octets,Gigawords,Packets}` attributes in RADIUS reporting and `ChilliSpot-Max-{Input,Output,Total}-{Octets,Gigawords}` for RADIUS-based traffic caps.

@PWJW @badnbusy @nemesifier feel free to chime in with your test results, thanks

Fixes: #9
Fixes: #25

f00b4r0 committed:
Implement uspot-bpf, an eBPF traffic counter

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

f00b4r0 committed:
uspot: client_interim(): remove dead code

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

f00b4r0 committed:
uspot: implement traffic limiting

Fixes: #25

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

f00b4r0 committed:
uspot: radius_acct(): don't needlessly call uspotfilter

We have all the necessary data internally already, this was kept as
bagage from previous implementation.

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

f00b4r0 committed:
introduce uspotbpf.uc

A wrapper for the eBPF traffic accounting module

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

f00b4r0 committed:
uspotfilter: cleanup unused API

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

f00b4r0 committed:
uspotbpf: work around ucode-mod-bpf bug

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

f00b4r0 committed:
uspot: implement swapio

This allows swapping the Input and Output side of the RADIUS attributes:

Acct-Input-*/Acct-Output-*
ChilliSpot-Max-Output-*/ChilliSpot-Max-Input-*

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

f00b4r0 committed:
handler-api: include 'seconds-remaining' only if not captive

Per RFC.

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

f00b4r0 committed:
handler-api: implement 'bytes-remaining'

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

f00b4r0 committed:
uspot: start(): ratelimit clearing is independent from RADIUS accounting

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

f00b4r0 committed:
README update

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

f00b4r0 committed:
uspotbpf: fix priority conflict with ratelimit

ratelimit and uspotbpf both used prio 512, stepping on each other's foot
set a higher priority for the bpf module otherwise it will not see the
ingress traffic stolen by the ifb interface.

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

f00b4r0 committed:
uspot: hook traffic accounting

Fixes: #9

Signed-off-by: Thibaut VARÈNE <hacks@slashdirt.org>

PWJW commented:
I think the "bytes_remaining" formatting is messed up on the `uspot clients` command:

```
hotspot:
{
        "92:2c:96:76:40:70": {
                "status": "Authenticated",
                "time": 501,
                "idle": 0,
                "lost": 0,
                "ip4addr": "10.1.0.236",
                "ip6addr": "",
                "sessionid": "3fc700efbec4baa5",
                "username": null,
                "client_ip": "10.1.0.236",
                "duration": 501,
                "packets_in": 83,
                "bytes_in": 11284,
                "packets_out": 67,
                "bytes_out": 10192,
                "seconds_remaining": 13899,
                "bytes_remaining": 102400000204778524
        }
}
```

The following was sent in the RADIUS Access-Accept response:

```
ChilliSpot-Max-Input-Octets=10240000
ChilliSpot-Max-Output-Octets=20480000
```

It looks like its not combining the bytes, but simply concatenating them together as a string?

Could `bytes_remaining` be separated for both rx and tx if different limits have been set? `bytes_remaining` is ambiguous in this use case, but okay if just `ChilliSpot-Max-Total-Octets` was used.

Thanks

f00b4r0 commented:
> I think the "bytes_remaining" formatting is messed up on the `uspot clients` command:
> It looks like its not combining the bytes, but simply concatenating them together as a string?

It looks like it indeed, it seems I've been bitten again by ucode's default type interpretation. I've just updated the branch and this should now be fixed.

> Could `bytes_remaining` be separated for both rx and tx if different limits have been set? `bytes_remaining` is ambiguous in this use case, but okay if just `ChilliSpot-Max-Total-Octets` was used.

Don't assume anything about  `bin/uspot`. In fact I already regret having mentioned it. This tool is primarily a convenience debug tool but nothing about it is set in stone, it's not even certain that the tool will be kept around.

`bytes_remaining` is computed by uspot internally for use in the captive portal API. The API does not distinguish between input and output bytes so there is no point in distinguishing there (other than for debugging purposes and I don't plan to add computations only for debugging purposes, so I'll see how to deal with that).

PWJW commented:
> > I think the "bytes_remaining" formatting is messed up on the `uspot clients` command:
> > It looks like its not combining the bytes, but simply concatenating them together as a string?
> 
> It looks like it indeed, it seems I've been bitten again by ucode's default type interpretation. I've just updated the branch and this should now be fixed.
> 
> > Could `bytes_remaining` be separated for both rx and tx if different limits have been set? `bytes_remaining` is ambiguous in this use case, but okay if just `ChilliSpot-Max-Total-Octets` was used.
> 
> Don't assume anything about `uspot`. In fact I already regret having mentioned it. This tool is primarily a convenience debug tool but nothing about it is set in stone, it's not even certain that the tool will be kept around.
> 
> `bytes_remaining` is computed by uspot internally for use in the captive portal API. The API does not distinguish between input and output bytes so there is no point in distinguishing there (other than for debugging purposes and I don't plan to add computations only for debugging purposes, so I'll see how to deal with that).

Understood re the tool, and thanks again for the work on this project - it's really appreciated. It's so much more modern and faster than CoovaChilli full stop 👍

f00b4r0 commented:
> > > I think the "bytes_remaining" formatting is messed up on the `uspot clients` command:
> > > It looks like its not combining the bytes, but simply concatenating them together as a string?
> > 
> > 
> > It looks like it indeed, it seems I've been bitten again by ucode's default type interpretation. I've just updated the branch and this should now be fixed.
> > > Could `bytes_remaining` be separated for both rx and tx if different limits have been set? `bytes_remaining` is ambiguous in this use case, but okay if just `ChilliSpot-Max-Total-Octets` was used.
> > 
> > 
> > Don't assume anything about `uspot`. In fact I already regret having mentioned it. This tool is primarily a convenience debug tool but nothing about it is set in stone, it's not even certain that the tool will be kept around.
> > `bytes_remaining` is computed by uspot internally for use in the captive portal API. The API does not distinguish between input and output bytes so there is no point in distinguishing there (other than for debugging purposes and I don't plan to add computations only for debugging purposes, so I'll see how to deal with that).
> 
> Understood re the tool, and thanks again for the work on this project - it's really appreciated. It's so much more modern and faster than CoovaChilli full stop 👍

No problem, thanks for the encouragement it's much appreciated, so is the testing. Let me know if I've fixed this issue because if the number shows up wrong in the tool, it means the daemon computed it wrong in the first place :)

f00b4r0 commented:
@PWJW please let me know if the bytes_remaining number looks sane now? thanks

PWJW commented:
@f00b4r0 My apologies, I've just re-tested this, and the limits are working, but I don't see the bytes_remaining showing in uspot clients anymore?

Here is the log and output:

```
Sat Jun 28 08:53:42 2025 daemon.debug uspot: hotspot 92:2c:96:76:40:70 creating client
Sat Jun 28 08:53:42 2025 daemon.info uspot: hotspot 92:2c:96:76:40:70 adding client
Sat Jun 28 08:53:42 2025 daemon.debug uspot: hotspot 92:2c:96:76:40:70 acct start
Sat Jun 28 08:53:42 2025 daemon.debug uspot: hotspot 92:2c:96:76:40:70 enabling quota limits: tx rx
[...perform some download...]
Sat Jun 28 08:55:03 2025 daemon.debug uspot: hotspot 92:2c:96:76:40:70 acct terminate: 5
Sat Jun 28 08:55:03 2025 daemon.info uspot: hotspot 92:2c:96:76:40:70 removing client: max dl octets reached
Sat Jun 28 08:55:03 2025 daemon.debug uspotfilter: hotspot 92:2c:96:76:40:70 client_remove
```

The RADIUS reply was:

```
Sat Jun 28 08:53:42 2025 : Debug: (1)   ChilliSpot-Max-Input-Octets := 10240000
Sat Jun 28 08:53:42 2025 : Debug: (1)   ChilliSpot-Max-Output-Octets := 20480000
Sat Jun 28 08:53:42 2025 : Debug: (1)   Acct-Interim-Interval = 240
```


```
# uspot clients
hotspot:
{
        "92:2c:96:76:40:70": {
                "status": "Authenticated",
                "time": 7,
                "idle": 0,
                "lost": 0,
                "ip4addr": "10.1.0.236",
                "ip6addr": "",
                "sessionid": "5a6d7bf1e93942e0",
                "username": null,
                "client_ip": "10.1.0.236",
                "duration": 7,
                "packets_in": 137,
                "bytes_in": 19467,
                "packets_out": 122,
                "bytes_out": 196224,
                "seconds_remaining": 14393
        }
}

```

f00b4r0 commented:
@PWJW I've pushed a minor change but I don't expect this is the cause of what you're seeing: I've eventually managed to get a testing setup and AFAICT it works here, yet I didn't touch the bytes_remaining part. Are you sure you don't have some leftovers of your manual edits in the code?

I'd suggest rebuilding the package from the current branch HEAD and reinstalling it cleanly. As far as I can tell the implementation is now correct but it certainly wouldn't hurt to have another confirmation :)

PWJW commented:
> @PWJW I've pushed a minor change but I don't expect this is the cause of what you're seeing: I've eventually managed to get a testing setup and AFAICT it works here, yet I didn't touch the bytes_remaining part. Are you sure you don't have some leftovers of your manual edits in the code?
> 
> I'd suggest rebuilding the package from the current branch HEAD and reinstalling it cleanly. As far as I can tell the implementation is now correct but it certainly wouldn't hurt to have another confirmation :)

I recompiled and can now see the bytes_remaining 👍 

```
Sat Jun 28 13:18:28 2025 daemon.debug uspot: hotspot 92:2c:96:76:40:70 creating client
Sat Jun 28 13:18:28 2025 daemon.info uspot: hotspot 92:2c:96:76:40:70 adding client
Sat Jun 28 13:18:28 2025 daemon.debug uspot: hotspot 92:2c:96:76:40:70 acct start
Sat Jun 28 13:18:28 2025 daemon.debug uspot: hotspot 92:2c:96:76:40:70 enabling quota limits on tx rx
[...download...]
Sat Jun 28 13:19:07 2025 daemon.debug uspot: hotspot 92:2c:96:76:40:70 acct terminate: 5
Sat Jun 28 13:19:07 2025 daemon.info uspot: hotspot 92:2c:96:76:40:70 removing client: max dl octets reached
Sat Jun 28 13:19:07 2025 daemon.debug uspotfilter: hotspot 92:2c:96:76:40:70 client_remove
```

```
# uspot clients
hotspot:
{
        "92:2c:96:76:40:70": {
                "status": "Authenticated",
                "time": 11,
                "idle": 0,
                "lost": 0,
                "ip4addr": "10.1.0.236",
                "ip6addr": "",
                "sessionid": "7c9ffd3c29faaf0c",
                "username": null,
                "client_ip": "10.1.0.236",
                "duration": 11,
                "packets_in": 244,
                "bytes_in": 30255,
                "packets_out": 212,
                "bytes_out": 219120,
                "seconds_remaining": 14389,
                "bytes_remaining": 30470625
        }
}
```

download...

```
# uspot clients
hotspot:
{
        "92:2c:96:76:40:70": {
                "status": "Authenticated",
                "time": 36,
                "idle": 0,
                "lost": 0,
                "ip4addr": "10.1.0.236",
                "ip6addr": "",
                "sessionid": "7c9ffd3c29faaf0c",
                "username": null,
                "client_ip": "10.1.0.236",
                "duration": 36,
                "packets_in": 8291,
                "bytes_in": 874292,
                "packets_out": 16350,
                "bytes_out": 24345993,
                "seconds_remaining": 14364,
                "bytes_remaining": 5499715
        }
}
```

once exceeded (counts in negative, expected, until the client is kicked)

```
# uspot clients
hotspot:
{
        "92:2c:96:76:40:70": {
                "status": "Authenticated",
                "time": 38,
                "idle": 0,
                "lost": 0,
                "ip4addr": "10.1.0.236",
                "ip6addr": "",
                "sessionid": "7c9ffd3c29faaf0c",
                "username": null,
                "client_ip": "10.1.0.236",
                "duration": 38,
                "packets_in": 14447,
                "bytes_in": 1312239,
                "packets_out": 30947,
                "bytes_out": 46742478,
                "seconds_remaining": 14362,
                "bytes_remaining": -17334717
        }
}
```

The figures look spot on!

f00b4r0 commented:
Thanks, this looks good. I'll merge in the coming days unless something else comes up :)

f00b4r0 commented:
@PWJW One last question: are the Input/Output limits implemented consistently with what you expect? (ie Output is client download, Input is upload, in both limits and accounting)

Thanks

f00b4r0 commented:
After receiving some feedback and looking at Coova's implementation of the "swapoctets" feature, I ended up renaming `acct_swapio` to `swapio` and adjusting the implementation so that it affects both accounting reports and quota limits altogether. Now the Acct-Input/Output attributes are consistent with the ChilliSpot-Max-Input/Output ones. (e.g. Input is always the same thing, up or download, in either case).

nemesifier commented on files/etc/config/uspot:
```diff
@@ -30,6 +30,8 @@
 #	option acct_proxy ''		# OPTIONAL - radius accounting server proxy
 #	option acct_interval ''		# OPTIONAL - radius accounting interim interval override
+#	option swapio ''		# OPTIONAL - swap the meaning of RADIUS 'Input' and 'Output' attributes
+#	option counters '0'		# OPTIONAL - enable traffic counters (for accounting and/or quota limits)
```
If this is set to 0, traffic accounting won’t work, is that correct?

The name of this option might be slightly misleading. Some users may want to track traffic for reporting purposes only, without enforcing counters or quota limits. Others may want to start with statistics and add limits later based on usage data.

In that context, setting it to 0 might seem appropriate, but it would unintentionally disable traffic reporting altogether.

Have you considered naming it something like `acct_traffic`, with a comment such as:
`# OPTIONAL – enable support for traffic accounting and quota limits`?

f00b4r0 replied:
> If this is set to 0, traffic accounting won’t work, is that correct?

Yes, as the explanation suggests. 0 is the default value.

> The name of this option might be slightly misleading. Some users may want to track traffic for reporting purposes only, without enforcing counters or quota limits. Others may want to start with statistics and add limits later based on usage data.

I'm not sure what's misleading here. Counters are counters. What you do with the counter values is up to you. This option governs accounting and traffic limiting (which are completely independent: you can have one without the other).

> In that context, setting it to 0 might seem appropriate,

Appropriate for what? I'm afraid I don't follow.

> but it would unintentionally disable traffic reporting altogether.
> 
> Have you considered naming it something like `acct_traffic`, with a comment such as: `# OPTIONAL – enable support for traffic accounting and quota limits`?

`acct_` prefixed options are *only* related to RADIUS accounting. For the same reason I renamed  `acct_swapio` to `swapio` I think such a prefix would be inadequate here.

nemesifier replied:
To clarify, more than prefixing "acct_" something I was trying to suggest to focus the naming and the comment on the traffic accounting feature rather than the quota limits.

I looked at the accounting sub section of the captive portal section of pfsense, one of the most widely used open source firewalls and captive portals out there:
 
<img width="1145" height="578" alt="Screenshot from 2025-07-17 18-38-40" src="https://github.com/user-attachments/assets/6273e54f-713b-44df-b1bc-86353f560837" />

Accounting style is equivalent to the swapio. I thought it could be a good idea to make the config options similar, but they don't have an option to turn off only traffic accounting. If accounting is turned on there, then the user gets both time and traffic, users can choose if they want to receive periodic accounting updates or only when the session ends, not sure if it's useful, just an observation I wanted to share.

nemesifier commented on files/etc/config/uspot:
```diff
@@ -30,6 +30,8 @@
 #	option acct_secret ''		# REQUIRED if acct_server is set - radius accounting server password
 #	option acct_proxy ''		# OPTIONAL - radius accounting server proxy
 #	option acct_interval ''		# OPTIONAL - radius accounting interim interval override
```
I assume this is in seconds right?

f00b4r0 replied:
Correct

nemesifier commented on files/etc/radcli/dictionary.chillispot:
```diff
@@ -9,6 +9,9 @@ ATTRIBUTE	ChilliSpot-Config			6	string	ChilliSpot
 ATTRIBUTE	ChilliSpot-Lang				7	string	ChilliSpot
 ATTRIBUTE	ChilliSpot-Version			8	string	ChilliSpot
 ATTRIBUTE	ChilliSpot-OriginalURL			9	string	ChilliSpot
+ATTRIBUTE	ChilliSpot-Max-Input-Gigawords		21	integer	ChilliSpot
```
While reviewing these attributes I noticed a discrepancy with the dictionary published on https://github.com/coova/coova-chilli/blob/master/doc/dictionary.coovachilli,

Here, the Vendor and attribtues are `ChilliSpot`, while in the coova-chilli repository it's  `CoovaChilli`, I am not sure if this is going to cause issues.
From what I know ChilliSpot is the old defunct project which was forked and renamed to CoovaChilli, which is still maintained although not actively developed.

PWJW replied:
Hi @nemesifier 

I'll answer some of the questions based on my findings:

The `counters 1` option does not enable or enforce any bandwidth or usage limits, it merely turns on the logging of data usage which is then included in any radius accounting packets. For bandwidth and/or usage limits to be applied to the client, you must include those specific radius attributes in the reply. If you don't, there is no limit applied.

The `acct_interval` is in seconds, correct.

For bandwidth rate limiting, I've been using this successfully with the following attributes:

```
WISPr-Bandwidth-Max-Down
WISPr-Bandwidth-Max-Up
```

this is in bit/s so to enforce a 1Mbit/s cap the value is 1000000

*OR*

```
ChilliSpot-Bandwidth-Max-Down
ChilliSpot-Bandwidth-Max-Up
```

this in kbit/s so to enforce a 1Mbit/s cap the value should be 1000.

Although CoovaChilli did take over the project from ChilliSpot, the RADIUS attributes were still named Chillispot-* for compatibility. These are what uspot uses so just choose from teh above two and run with it 👍 

Thanks for testing!

nemesifier replied:
The radius attributes are numbers, so it doesn't matter much if it's prefixed as ChilliSpot or CoovaChilli, as long as the choice is consistent in the dictionary and attributes used in the application.

f00b4r0 replied:
@nemesifier you answered your own question above so all is good. CoovaChilli renamed the attributes in their own dictionary, they're welcome to do that but the Vendor Name that was officially registered is ChilliSpot.

As a matter of fact that's what freeRADIUS uses: https://github.com/FreeRADIUS/freeradius-server/blob/master/share/dictionary/radius/dictionary.chillispot

nemesifier reviewed (commented):
Great work @f00b4r0! 

I have been testing this for the last few days, here's what I have observed up to now:

- Traffic accounting works after enabling the "counters" option, which seems to be expected. :clap: 
- Swapio works as expected. :clap: 
- Traffic quota works. :clap: 
- I couldn't get rate limiting to work, I will do more testing, I may be doing something wrong with the attribute names (I left a comment below regarding the name of the attributes), I will review and come back on this.
 
During initial testing, this log line came up:

```
Sat Jul 12 20:35:38 2025 kern.warn kernel: [63374.666485] uspot.uc[5680]: memfd_create() called without MFD_EXEC or MFD_NOEXEC_SEAL set
```

But didn't come up later anymore.

My next steps:

1. Test gigawords RADIUS attributes
2. Test again rate limiting

f00b4r0 commented:
> Great work @f00b4r0!
> 
> I have been testing this for the last few days, here's what I have observed up to now:
> 
> * Traffic accounting works after enabling the "counters" option, which seems to be expected. 👏
> * Swapio works as expected. 👏
> * Traffic quota works. 👏

Good, I'll merge now.

> * I couldn't get rate limiting to work, I will do more testing, I may be doing something wrong with the attribute names (I left a comment below regarding the name of the attributes), I will review and come back on this.

Have you installed the ratelimit package? Should work out of the box.
> 
> During initial testing, this log line came up:
> 
> ```
> Sat Jul 12 20:35:38 2025 kern.warn kernel: [63374.666485] uspot.uc[5680]: memfd_create() called without MFD_EXEC or MFD_NOEXEC_SEAL set
> ```

AFAICT it's harmless. I'll probably report it against the ucode repository if I can come up with a small reproducer.

nemesifier commented:
I just wanted to report that I also tested the gigawords attribute `ChilliSpot-Max-Total-Gigawords` and works as expected, thanks :+1:
